### DISCUSSION TITLE
rare System.IndexOutOfRangeException when reusing EncodedConnection object

### DISCUSSION THREAD

**Discussion by shimr12** [nats-io/nats.net.v1#713] [Announcements]
Hi.
I use a very small pool of `EncodedConnection`s and share them between threads.
I saw [here](https://github.com/nats-io/nats.net/issues/350) a mention that `EncodedConnection` is thread safe so my usage is normal.

I rarely get a `System.IndexOutOfRangeException` when trying to call `EncodedConnection.SubscribeAsync`.

example stacktrace:
```
System.IndexOutOfRangeException: Index was outside the bounds of the array.\n   
at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)\n   
at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)\n   
at NATS.Client.EncodedConnection.subscribeAsync(String subject, String reply, EventHandler`1 handler)\n   
at NATS.Client.EncodedConnection.SubscribeAsync(String subject, String queue, EventHandler`1 handler)\n   
```

calling `Dictionary.Add` shouldn't raise `System.IndexOutOfRangeException` normally. 
usually it means your using it from multiple threads without locking.

I took a glance at the code [here](https://github.com/nats-io/nats.net/blob/59eca379ead07610244a385b4c110ce7987c13f3/src/NATS.Client/EncodedConn.cs#L279) and it confirms it.

if `EncodedConnection` supposed to be thread-safe this is a bug.

**Comment by hastarin** [comment]
Perhaps a ConcurrentDictionary should be used instead. 
https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=netframework-4.6.2
